src/main/java/mods/eln/sixnode/EmergencyLamp.kt

```diff
@@ -0,0 +1,320 @@
+                }
+            else
+                preserveMatrix {
+                    mainWall.draw()
+                    (if (mirrorSign) mainWallL else mainWallR).draw()
+                }
+        }
+    }
+
+    override fun getFrontFromPlace(side: Direction?, player: EntityPlayer?)
+        = super.getFrontFromPlace(side, player).inverse()
+
+    override fun canBePlacedOnSide(player: EntityPlayer?, side: Direction?): Boolean {
+        if (side == Direction.YN) {
+            Utils.addChatMessage(player, I18N.tr("Emergency lamp can not be placed on ground."))
+            return false
+        } else
+            return true
+    }
+
+    override fun addInformation(itemStack: ItemStack?, entityPlayer: EntityPlayer?, list: MutableList<String>,
+                                par4: Boolean) {
+        with(list) {
+            add("As long as power is provided, the internal battery")
+            add("is charged and the lamp is off. On a power failure,")
+            add("the lamp turns on and runs on batteries.")
+            add(Utils.plotVolt("Nominal voltage:", cable.electricalNominalVoltage))
+            add(Utils.plotEnergy("Battery capacity:", batteryCapacity))
+        }
+    }
+}
+
+class EmergencyLampElement(sixNode: SixNode, side: Direction, descriptor: SixNodeDescriptor)
+    : SixNodeElement(sixNode, side, descriptor) {
+
+    enum class Event(val value: Byte) {
+        TOGGLE_POWERED_BY_CABLE(1),
+        SET_CHANNEL(2)
+    }
+
+    val desc = descriptor as EmergencyLampDescriptor
+    val load = NbtElectricalLoad("load")
+    val chargingResistor = ResistorSwitch("chargingResistor", load, null)
+    var on = false
+        set(value) {
+            if (field != value) needPublish()
+            field = value
+            sixNode.lightValue = if (value) desc.lightLevel else 0
+        }
+    var charge = desc.batteryCapacity / 2
+    var poweredByCable = false
+        set(value) {
+            val changed = field != value
+            field = value
+            if (changed) {
+                reconnect()
+                needPublish()
+                if (value) isConnectedToLampSupply = false
+            }
+        }
+
+    var channel by published("Default channel")
+    var isConnectedToLampSupply by published(false)
+
+    val process = IProcess { deltaT ->
+        if (!poweredByCable) {
+            var closestPowerSupply: LampSupplyElement.PowerSupplyChannelHandle? = null
+            var closestDistance = 10000f
+
+            LampSupplyElement.channelMap[channel]?.forEach {
+                val distance = it.element.sixNode.coordonate.trueDistanceTo(sixNode.coordonate).toFloat()
+                if (distance < closestDistance && distance <= it.element.range) {
+                    closestDistance = distance
+                    closestPowerSupply = it
+                }
+            }
+
+            if (closestPowerSupply != null && closestPowerSupply!!.element.getChannelState(closestPowerSupply!!.id)) {
```

Comment:
You shouldn't need the !!s, given the explicit null test. (And IDEA should be telling you that.)